Provide server-side meter (#4972)
apache/skywalking: docs/en/setup/backend/backend-meter.md

# Meter Receiver
Meter receiver is accepting the metrics of [meter protocol](https://github.com/apache/skywalking-data-collect-protocol/blob/master/language-agent/Meter.proto) format into the [Meter System](./../../concepts-and-designs/meter.md).

## Module define
receiver-meter:
  selector: ${SW_RECEIVER_METER:default}
  default:

## Configuration file
Meter receiver is configured via a configuration file. The configuration file defines everything related to receiving 
 from agents, as well as which rule files to load.
 
OAP can load the configuration at bootstrap. If the new configuration is not well-formed, OAP fails to start up. The files
are located at `$CLASSPATH/meter-receive-config`.

The file is written in YAML format, defined by the scheme described below. Brackets indicate that a parameter is optional.

> A example can be found [here](../../../../oap-server/server-bootstrap/src/main/resources/meter-receive-config/spring.yaml)

Reviewer: And `backend-setup.md` should add a link of `Spring Sleuth Metrics Analysis` section in `Advanced feature document link list`

> A example can be found [here](../../../../oap-server/server-bootstrap/src/main/resources/meter-receive-config/spring.yaml)

Reviewer: `Spring Sleuth Metrics Analysis` provides a fully detailed documents including 
1. Set up agent
1. Set up backend receiver
1. Add UI dashboard 

And the screenshot could be hosted in `skywalking-website` repo, there are a lot of screenshots hosted there.

- `subtract`. Subtract value into meter. Support single value.
> meter["test_meter"].subtract(2)

In this case, all of the meter values will subtract `2`.
> meter["test_meter1"].subtract(meter["test_meter2"])

In this case, all of the `test_meter1` values will subtract value from `test_meter2`, ensure `test_meter2` only has single value to operate, could use `tagFilter`.
- `multiply`. Multiply value into meter. Support single value.
> meter["test_meter"].multiply(2)

In this case, all of the meter values will multiply `2`.
> meter["test_meter1"].multiply(meter["test_meter2"])

In this case, all of the `test_meter1` values will multiply value from `test_meter2`, ensure `test_meter2` only has single value to operate, could use `tagFilter`.
- `divide`. Divide value into meter. Support single value.
> meter["test_meter"].divide(2)

In this case, all of the meter values will divide `2`.
> meter["test_meter1"].divide(meter["test_meter2"])

In this case, all of the `test_meter1` values will divide value from `test_meter2`, ensure `test_meter2` only has single value to operate, could use `tagFilter`.

> - `scale`. Scale value into meter. Support single value.

Author: Has add a example.

# Meter Receiver
Meter receiver is accept meter from [meter protocol](https://github.com/apache/skywalking-data-collect-protocol/blob/master/language-agent/Meter.proto) into the [Meter System](./../../concepts-and-designs/meter.md).o

## Module define
receiver-meter:
  selector: ${SW_RECEIVER_METER:default}
  default:

## Configuration file
Meter receiver is configured via a configuration file. The configuration file defines everything related to receiving 
 from agents, as well as which rule files to load.
 
OAP can load the configuration at bootstrap. If the new configuration is not well-formed, OAP fails to start up. The files

> are located at `$CLASSPATH/meter-receive-config`.

Reviewer: I think you should provide Sleuth config in the folder, the example of agent setup, UI template, and full documentation to explain the document.
There is nothing than an example to show how to use this feature.

The file is written in YAML format, defined by the scheme described below. Brackets indicate that a parameter is optional.

> A example can be found [here](../../../../oap-server/server-bootstrap/src/main/resources/meter-receive-config/meter-receive-config.yaml)

Reviewer: This file is not available. Also at the same time, what is this file for?

Reviewer: Do you plan to have a config for Spring app?

Author: Rename to the `spring.yaml`, support moniting on the spring application.

### Meters configure

```yaml
# Meter config allow your to recompute
meters:
  # Meter name which combines with a prefix 'meter_' as the index/table name in storage.
  - name: <string>
    # The meter scope
    scope:
      # Scope type should be one of SERVICE, SERVICE_INSTANCE, ENDPOINT
      type: <string>
      # Appoint the endpoint name if using ENDPOINT scope
      endpoint: <string>
    # The agent source of the transformation operation.
    meter:
      # The transformation operation from prometheus metrics to Skywalking ones. 
      operation: <string>
      # Meter value parse groovy script.
      value: <string>

>       # Appoint percentiles if using avgHistogramPercentile operation.

Reviewer: ```suggestion
      # <Optional> Appoint percentiles if using avgHistogramPercentile operation.
```

      percentile:
        - <rank>
```

#### Meter transform operation

The available operations are `avg`, `avgHistogram` and `avgHistogramPercentile`. The `avg` and `avgXXX` mean to average
the raw received metrics. 

When you specify `avgHistogram` and `avgHistogramPercentile`, the source should be the type of `histogram`.

#### Meter value script

The script is provide a easy way to custom build a complex value, and it also support combine multiple meter into one.

> Use `meter[""]` to refer the metrics raw data, and multiple build-in methods to help filter or operate the value.

Reviewer: We need grammar expressions. Take a look on SkyWalking's OAL doc and grammar expressions.

Author: Has changed the doc, add the grammar expressions explain.

> Using `meter[""]` to get one meter from all of the meter witch received at ***per agent***, also there have multiple build-in method to help filter or operate the value.

Reviewer: ```suggestion
Use `meter[""]` to refer the metrics raw data, and multiple build-in methods to help filter or operate the value.
```

> It will combine the values and collect into the Meter System.

Reviewer: ```suggestion

```

Reviewer: Don't use `future tense` if possible.

|Function|Support meter type|Description|
|------|-------|------|
|tagFilter(String, String)|Counter,Gauge,Histogram|Filter tag key/value from meter|

> |add(double)|Counter,Gauge|Add value into meter|

Reviewer: Do we have counter and guage type? In the protocol, we define https://github.com/apache/skywalking-data-collect-protocol/blob/master/language-agent/Meter.proto#L49. There is no such thing. You must keep the nouns consistent, otherwise, people are going to be very confused. Even I am hard to follow.

Author: Have changed to the `SingleValue` or `Histogram`

|add(Meter)|Counter,Gauge|Add value from a meter value|
|reduce(double)|Counter,Gauge|Reduce value into meter|
|reduce(Meter)|Counter,Gauge|Reduce value from a meter value|
|multiply(double)|Counter,Gauge|Multiply value into meter|
|multiply(Meter)|Counter,Gauge|Multiply value form a meter value|
|mean(double)|Counter,Gauge|Mean value into meter|

> |mean(Meter)|Counter,Gauge|Mean value from a meter value|

Reviewer: What do you mean `mean`? A little confused about this term

|scale(int)|Counter,Gauge|Scale the meter value|
|rate(string)|Counter,Gauge,Histogram|Rate value from the time range|
|irate(string)|Counter,Gauge,Histogram|IRate value from the time range|
|increase(string)|Counter,Gauge,Histogram|Increase value from the time range|

> The `add`, `reduce`, `multiply`, `mean` with Meter only support operate from single Meter value, you could using `tagFilter` to filter, Because we could not operation between two multiple values.

Reviewer: ```suggestion
The `add`, `reduce`, `multiply`, and `mean` support single Meter value. Use `tagFilter` to filter.
```

What do you mean `Because we could not operation between two multiple values.`?

> If you want to using `rate`, `irate`, `increase` function, I suggest to let it rate counter at the agent side. 

Reviewer: ```suggestion
If you want to use `rate`, `irate`, `increase` function, use client-side API.
```

> Because if the connection between agent and backend has reconnected, the time windows will be break, so we cannot find previous value from the window.

Reviewer: ```suggestion

- FAQ, why no `rate`, `irate`, `increase` at the backend.
Once the agent reconnected to another OAP instance, the time windows of rate calculation will break. Then, the result would not be accurate.
```
